File: verifiers/web/conformance.mjs
// Run the browser verification core against the normative vectors under
// Node (same module the page uses). Output format matches the Go verifier
// so scripts/conformance.py can build the cross-implementation table.
//
// Usage: node conformance.mjs [vector-dir]
import { readFileSync } from "node:fs";
import { createPrivateKey, createPublicKey } from "node:crypto";
import {
  anchorInputDigest, b64uEncode, canonicalize, digestJson, digestBytes,
  eventHash, signingBytes, verifyChain, verifySegmentClose,
  verifySegmentSequence, verifySigned,
} from "./verifier-core.mjs";

const dir = process.argv[2] ?? "../../spec/vectors/v0.1";
const vectors = JSON.parse(readFileSync(`${dir}/vectors.json`, "utf8"));
const keysDoc = JSON.parse(readFileSync(`${dir}/keys.json`, "utf8"));

// Derive raw public keys from the vector seeds (keys.json contract) using
// node:crypto only for the seed->public derivation.
const keys = new Map();
for (const entry of Object.values(keysDoc.keys)) {
  const seed = Buffer.from(entry.seed_hex, "hex");
  const der = Buffer.concat([
    Buffer.from("302e020100300506032b657004220420", "hex"), seed]);
  const pub = createPublicKey(createPrivateKey({ key: der, format: "der", type: "pkcs8" }))
    .export({ format: "der", type: "spki" });
  const raw = new Uint8Array(pub.subarray(pub.length - 32));
  if (b64uEncode(raw) !== entry.public_key_b64u) {
    throw new Error(`keys.json inconsistent for ${entry.key_id}`);
  }
  keys.set(entry.key_id, raw);
}

async function expectVerdict(fn, expected) {
  let verified = true, err = null;
  try { await fn(); } catch (e) { verified = false; err = e; }
  if (verified !== expected) {
    throw new Error(`verified=${verified} want ${expected} (${err?.message ?? ""})`);
  }
}

async function runCase(c) {
  const exp = c.expected;
  switch (c.kind) {
    case "sign-verify": {
      await expectVerdict(
        () => verifySigned(c.object, c.type, keys.get(c.verify_key)), exp.verified);
      if (exp.verified && exp.canonical_signing_object_b64u) {
        const canonical = signingBytes(c.object, "");
        if (b64uEncode(canonical) !== exp.canonical_signing_object_b64u) {
          throw new Error("canonical signing bytes differ");
        }
      }
      return;
    }
    case "event-hash": {
      if ((await eventHash(c.object)) !== exp.event_hash) {
        throw new Error("event_hash differs");
      }
      if (exp.canonical_b64u &&
          b64uEncode(canonicalize(c.object)) !== exp.canonical_b64u) {
        throw new Error("canonical bytes differ");
      }
      return;
    }
    case "chain":
      return expectVerdict(() => verifyChain(c.events), exp.verified);
    case "segment":
      return expectVerdict(
        () => verifySegmentClose(c.close, c.writer_auth, c.events, keys), exp.verified);
    case "segment-chain":
      return expectVerdict(() => verifySegmentSequence(c.closes), exp.verified);
    case "anchor": {
      if ((await anchorInputDigest(c.object)) !== exp.digest) {
        throw new Error("anchor input digest differs");
      }
      return;
    }
    case "digest-json": {
      if ((await digestJson(c.value)) !== exp.digest) throw new Error("digest_json differs");
      return;
    }
    case "digest-bytes": {
      const raw = Uint8Array.from(Buffer.from(c.value_hex, "hex"));
      if ((await digestBytes(raw)) !== exp.digest) throw new Error("digest_bytes differs");
      return;
    }
    default:
      throw new Error(`unknown kind ${c.kind}`);
  }
}

let failures = 0;
for (const c of vectors.cases) {
  try {
    await runCase(c);
    console.log(`${c.name}: PASS`);
  } catch (e) {
    console.log(`${c.name}: FAIL ${e.message}`);
    failures++;
  }
}
console.log(`TOTAL: ${vectors.cases.length - failures}/${vectors.cases.length}`);
process.exit(failures ? 1 : 0);

File: scripts/conformance.py
"""Cross-implementation conformance table: Python (reference SDK) vs Go
(verifiers/go, implemented from the specification alone).

Usage: python scripts/conformance.py <go-output-file>
where the file holds the Go verifier's per-case output lines.
Prints a markdown table and exits non-zero unless BOTH columns pass on
every case.
"""

import json
import sys
from pathlib import Path

from oathon import crypto
from oathon.crypto import VerificationError
from oathon.verify import (
    KeySet,
    verify_chain,
    verify_segment_close,
    verify_segment_sequence,
)

ROOT = Path(__file__).resolve().parent.parent
VDIR = ROOT / "spec" / "vectors" / "v0.1"


def python_result(case, keyset, raw_keys):
    kind = case["kind"]
    expected = case["expected"]
    try:
        if kind == "sign-verify":
            pub = raw_keys[case["verify_key"]]
            try:
                crypto.verify_object(case["object"],
                                     crypto.SIGN_DOMAINS[case["type"]], pub)
                verified = True
            except VerificationError:
                verified = False
            if verified != expected["verified"]:
                return False
            if verified and "canonical_signing_object_b64u" in expected:
                canonical = crypto.b64u_encode(crypto.signing_input(case["object"], b""))
                if canonical != expected["canonical_signing_object_b64u"]:
                    return False
            return True
        if kind == "event-hash":
            if crypto.event_hash(case["object"]) != expected["event_hash"]:
                return False
            if "canonical_b64u" in expected:
                return crypto.b64u_encode(
                    crypto.canonicalize(case["object"])) == expected["canonical_b64u"]
            return True
        if kind == "chain":
            try:
                verify_chain(case["events"])
                return expected["verified"] is True
            except VerificationError:
                return expected["verified"] is False
        if kind == "segment":
            try:
                verify_segment_close(case["close"], case["events"],
                                     case["writer_auth"], keyset)
                return expected["verified"] is True
            except VerificationError:
                return expected["verified"] is False
        if kind == "segment-chain":
            try:
                verify_segment_sequence(case["closes"])
                return expected["verified"] is True
            except VerificationError:
                return expected["verified"] is False
        if kind == "anchor":
            return crypto.anchor_input_digest(case["object"]) == expected["digest"]
        if kind == "digest-json":
            return crypto.digest_json(case["value"]) == expected["digest"]
        if kind == "digest-bytes":
            return crypto.digest_bytes(
                bytes.fromhex(case["value_hex"])) == expected["digest"]
    except Exception:  # noqa: BLE001 — any crash is a conformance failure
        return False
    return False


def main() -> int:
    vectors = json.loads((VDIR / "vectors.json").read_text())
    keys = json.loads((VDIR / "keys.json").read_text())
    keyset = KeySet.from_json(
        {e["key_id"]: e["public_key_b64u"] for e in keys["keys"].values()})
    raw_keys = {e["key_id"]: crypto.b64u_decode(e["public_key_b64u"])
                for e in keys["keys"].values()}

    go_lines = Path(sys.argv[1]).read_text().splitlines()
    go_status = {}
    for line in go_lines:
        if ": " in line and not line.startswith("TOTAL"):
            name, status = line.split(": ", 1)
            go_status[name] = "PASS" if status.strip() == "PASS" else "FAIL"

    rows, ok = [], True
    for case in vectors["cases"]:
        py = "PASS" if python_result(case, keyset, raw_keys) else "FAIL"
        go = go_status.get(case["name"], "MISSING")
        rows.append((case["name"], py, go))
        ok = ok and py == "PASS" and go == "PASS"

    width = max(len(r[0]) for r in rows)
    print(f"| {'vector id'.ljust(width)} | Python | Go   |")
    print(f"|{'-' * (width + 2)}|--------|------|")
    for name, py, go in rows:
        print(f"| {name.ljust(width)} | {py.ljust(6)} | {go.ljust(4)} |")
    print(f"\n{len(rows)} vectors; Python "
          f"{sum(1 for r in rows if r[1] == 'PASS')}/{len(rows)}, Go "
          f"{sum(1 for r in rows if r[2] == 'PASS')}/{len(rows)}")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())
